blocks/algorithms.py

```diff
@@ -0,0 +1,227 @@
+"""Training algorithms."""
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+
+import theano
+from theano import tensor
+
+from blocks.utils import update_instance
+from blocks.graph import ComputationGraph
+
+
+class TrainingAlgorithm(object):
+    """Base class for training algorithms.
+
+    A training algorithm object has a simple life-cycle.
+    First it is initialized by calling its :meth:`initialize` method.
+    At this stage, for instance, Theano functions can be compiled.
+    After that the :meth:`process_batch` method is repeatedly
+    called with a batch of training data as a parameter.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def initialize(self):
+        """Initialize the training algorithm."""
+        pass
+
+    @abstractmethod
+    def process_batch(self, batch):
+        """Process a batch of training data.
+
+        Attributes
+        ----------
+        batch : dict
+            A dictionary of (source name, data) pairs.
+
+        """
+        pass
+
+
+class TheanoTrainingAlgorithm(TrainingAlgorithm):
+    """A training algorithm based on Theano functions.
+
+    Parameters
+    ----------
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+       Ideally the parameters should be extracted from the `cost`
+       computation graph. But in order for this to work our graph
+       traversal routine should recurse into ScanOp's, which is currently
+       not the case.
```

Comment:
I fully support that we need a good default implementation for the case 
of differentiable cost. The question discussed is: how to achieve this 
goal. My point is: let the training algorithm decide what it wants to do 
with computation graph of the cost given to it, with no middlemen, be it 
the main loop or a "model" involved.

I have already added a stupid automatic fetching of parameters to the 
`TheanoTrainingAlgorithm` class.

On 16.01.2015 11:17, Jan Chorowski wrote:
>
> In blocks/algorithms.py 
> <https://github.com/bartvm/blocks/pull/92#discussion-diff-23072225>:
>
> > +    cost : Theano variable
> > +        The objective to be minimized.
> > +    params : list of Theano shared variables
> > +        The parameters to be tuned.
> > +
> > +    Notes
> > +    -----
> > +        Changing `updates` attribute or calling `add_updates` after
> > +        the `initialize` method is called will have no effect.
> > +
> > +    .. todo::
> > +
> > +       Ideally the parameters should be extracted from the `cost`
> > +       computation graph. But in order for this to work our graph
> > +       traversal routine should recurse into ScanOp's, which is currently
> > +       not the case.
> Hey Dima, I don't think I understood you correctly. We are considering 
> an extremely popular use case here - how to train a model that is 
> differentiable wrt its parameters. Even if there are use cases that 
> fall outside of it we shouldn't stop supporting the common use case. 
> Unless we will provide a good default implementation, we will end up 
> with is a magical model setup code that everyone will end up 
> copy-pasting. I think it is better to have a generic main loop and 
> trainer and a Model/Trainable/Experiment class that is interfaced with 
> them. No one is forced to use it. It is not a problem if someone uses 
> just part of blocks. But all parts should be there for the common case.
> …
> On 1/16/2015 9:25 AM, Dmitry Bogdanov wrote: In blocks/algorithms.py 
> <https://github.com/bartvm/blocks/pull/92#discussion-diff-23067722>: > 
> + cost : Theano variable > + The objective to be minimized. > + params 
> : list of Theano shared variables > + The parameters to be tuned. > + 
> > + Notes > + ----- > + Changing `updates` attribute or calling 
> `add_updates` after > + the `initialize` method is called will have no 
> effect. > + > + .. todo:: > + > + Ideally the parameters should be 
> extracted from the `cost` > + computation graph. But in order for this 
> to work our graph > + traversal routine should recurse into ScanOp's, 
> which is currently > + not the case. I am not a big fan of the 
> solution proposed. First of all the name |Model| does not mean 
> anything. It was one the most obscure concepts in both PyLearn2 and 
> Groundhog. I did use the word model in the main loop class, but there 
> it has very limited semantics of a handle to everything we want save 
> for this experiment, with no assumption that a model has gradients or 
> a Theano expression for cost or whatever else. So I guess that I can 
> further discuss the proposal above using the name |Trainable| instead 
> of the name |Model|. Now we come to my main point: I believe there is 
> no such thing as /the/ trainable. There are just no traits shared by 
> all possible things that can be trained. I am looking forward to 
> repeating the experiments by Google DeepMind people with a 
> non-differentiable cost optimized by reinforcement learning: there are 
> no gradients in this case. It is harder to point at concrete examples 
> when some of my parameters are not Theano variables, but here comes 
> another consideration: it is quite weird when the cost and the 
> gradients are given by the user and parameters are provided by the 
> main loop. This is one of PyLearn2's shortcoming: different components 
> ask each other for different things and one struggles to understand 
> the information flow. So what I had in mind writing this code is: the 
> user is fully responsible for configuring the algorithm. No obscure 
> information passing through the main loop. And when it comes to a 
> particular algorithm or to a family of them, this is where making 
> assumptions is possible. And a base class of the family of algorithms, 
> call it |DifferentiableCostAlgorithm| working with a differentiable 
> Theano cost can provide convenience routines of collecting the 
> parameters, because at this point it is quite reasonable to assume 
> that those are shared Theano variables. And this actually makes the 
> |DifferentiableCostAlgorithm| usable without the main loop. I can just 
> do algo= GradientDescent(cost) for iin range(10): 
> algo.process_batch(some_data) without creating a |Model| or a 
> |Trainable| instance. And I see no problem whatsoever that the 
> |ComputationGraph| is needed for that, because there should be a 
> lowest common denominator between all library components, and I do not 
> care whether it is just Theano or Theano with our extension, such as 
> the |ComputationGraph| is. — Reply to this email directly or view it 
> on GitHub <https://github.com/bartvm/blocks/pull/92/files#r23067722>.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/92/files#r23072225>.
>